### PR TITLE
Metalline

### PR DESCRIPTION
**Pull request by chonma0ctopus**
This makes it possible to model atomic absorption lines based on the line lists by VALD (exojax.spec.moldb.AdbVald) or Kurucz (exojax.spec.moldb.AdbKurucz).
The line list data needs to be downloaded manually from the [VALD3](http://vald.astro.uu.se/) or [Kurucz](http://kurucz.harvard.edu/linelists/) webpages.

### CODE REVIEW

**Review by HajimeKawahara — COMMENTED** [review]
@chonma0ctopus Thank you for sending the PR! I started to review from performing pytest and tutorial. But I have a problem the dataset from VALD3. Can you show us some instruction how we should download the data from VALD3 for these tests. i.e. wavelength or metal choice and so on.
 

valddl

File: src/exojax/spec/atomll.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+       elower: elower
+       QTref_284: partition function Q(Tref)
+       QTmask: mask to identify a rows of QTref_284 to apply for each line
+       Irwin: if True(1), the partition functions of Irwin1981 is used, otherwise those of Barklem&Collet2016
+
+    Returns:
+       Sij(T): Line strength (cm)
+
+    """
+    ccgs=29979245800.0 #[cm/s]
```
Comment (HajimeKawahara) [inline src/exojax/spec/atomll.py:24]: We can import physical constants from [utils.constants](https://github.com/HajimeKawahara/exojax/blob/develop/src/exojax/utils/constants.py).

```python
from exojax.utils.constants import ccgs, hcperk
```
you can add other constants if you do not find them in exojax.utils.constants.

File: src/exojax/spec/atomll.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+      Barklem+2000: https://ui.adsabs.harvard.edu/abs/2000A&AS..142..467B
+      Gray+2005: https://ui.adsabs.harvard.edu/abs/2005oasp.book.....G
+    """
+    ccgs = 2.99792458e10 #[cm/s]
+    kcgs = 1.38064852e-16 #[erg/K]
+    gamRad = np.where(gamRad==0., -99, gamRad)
+    gamSta = np.where(gamSta==0., -99, gamSta)
+    hcgs = 6.62607015e-27 #Planck constant [erg*s]
+    Rcgs = 1.0973731568e5 #Rydberg constant [cm-1]
+    ucgs = 1.660539067e-24 #unified atomic mass unit [g]
```
Comment (HajimeKawahara) [inline src/exojax/spec/atomll.py:231]: ucgs = m_u in [utils.constants](https://github.com/HajimeKawahara/exojax/blob/develop/src/exojax/utils/constants.py).

File: src/exojax/spec/atomll.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+      Kurucz+1981: https://ui.adsabs.harvard.edu/abs/1981SAOSR.391.....K
+      Barklem+1998: https://ui.adsabs.harvard.edu/abs/1998MNRAS.300..863B
+      Barklem+2000: https://ui.adsabs.harvard.edu/abs/2000A&AS..142..467B
+      Gray+2005: https://ui.adsabs.harvard.edu/abs/2005oasp.book.....G
+    """
+    ccgs = 2.99792458e10 #[cm/s]
+    kcgs = 1.38064852e-16 #[erg/K]
+    gamRad = np.where(gamRad==0., -99, gamRad)
+    gamSta = np.where(gamSta==0., -99, gamSta)
+    chi_lam = nu_lines/8065.54 #[cm-1] -> [eV]
```
Comment (HajimeKawahara) [inline src/exojax/spec/atomll.py:92]: I found many 8065.54 in this file. Can you define it in  [utils.constants](https://github.com/HajimeKawahara/exojax/blob/develop/src/exojax/utils/constants.py) and use it?

File: src/exojax/spec/exomolapi.py [hunk 4]
```diff
@@ -214,7 +214,7 @@ def pickup_gEslow(states,trans):
     A=trans["A"].to_numpy()
     nu_lines=trans["nu_lines"].to_numpy()
     elower=trans["elower"].to_numpy()
-    gpu=trans["gup"].to_numpy()
+    gup=trans["gup"].to_numpy()
```
Comment (HajimeKawahara) [inline src/exojax/spec/exomolapi.py:217]: Thank you!

File: tests/opacity_Fe_test.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
+"""test for opacity of metal lines
```
Comment (HajimeKawahara) [inline tests/opacity_Fe_test.py:1]: Can you move these tests to tests/xs/ ?

File: src/exojax/spec/moldb.py [hunk 6]
```diff
@@ -623,3 +623,496 @@ def search_molecid(molec):
...
+        
+        """
+        #if len(atomspecies.split(' '))==2:
+        atomspecies_Roman = atomspecies.split(' ')[0] + '_' + 'I'*int(atomspecies.split(' ')[-1])
+        gQT = self.gQT_284species[ np.where(self.pfdat['T[K]']==atomspecies_Roman) ][0]
+        return gQT
+    
+    
+    
+    def QT_interp(self, atomspecies, T, Irwin=False):
```
Comment (HajimeKawahara) [inline src/exojax/spec/moldb.py:1041]: We will use this funtion in an HMC. ”if” statement is incompatible with JIT/XLA. Consider separating them into multiple functions or other solutions.

**Comment by chonma0ctopus** [comment]
@HajimeKawahara 
Thank you very much for the useful suggestions. I believe I've incorporated all of them. Could you confirm them?

File: tests/xs/opacity_Fe_test_gamma_uns.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+"""
+
+import pytest
+import numpy as np
+from exojax.spec import xsection, moldb, atomll
+from exojax.spec.hitran import SijT, doppler_sigma
+import matplotlib.pyplot as plt
+from exojax.utils.constants import m_u
+
+filepath_VALD3 = '/home/tako/work/VALD3/vald2600.gz'
```
Comment (HajimeKawahara) [inline tests/xs/opacity_Fe_test_gamma_uns.py:26]: Can you change this path to .database?

File: tests/xs/opacity_Fe_test_gamma_uns.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+path_fig = '/home/tako/Dropbox/tmpfig/tmp_211111/'
```
Comment (HajimeKawahara) [inline tests/xs/opacity_Fe_test_gamma_uns.py:27]: should be replaced to the current path ("./")?

File: tests/xs/opacity_Fe_test_gamma_uns.py (hunk 7 above)
Comment (HajimeKawahara) [reply tests/xs/opacity_Fe_test_gamma_uns.py:26]: Note that exojax does not assume a common path to the database. If not specified exojax searches for the database in the local .database directory.

File: tests/xs/opacity_Fe_test_gamma_KA4.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+#$ cp [user_name_at_VALD].[request_number_at_VALD].gz vald2600.gz
+adbFe = moldb.AdbVald(filepath_VALD3, nus4LL, Irwin=pf_Irwin)
+
+Amol=np.float64( adbFe.atomicmass[0] ) #atomic mass [u]
+ionE=np.float64( adbFe.ionE[0] ) #ionization energy [eV]
+nu0=adbFe.nu_lines
+
+#-------
+
+@pytest.mark.parametrize("T", [81, 200, 493, 1215, 2217, 2995, 3250, 4000]) #[81, 110, 148, 200, 270, 365, 493, 666, 900, 1215, 1641, 2000, 2217, 2500, 2750, 2995, 3250, 3500, 3750, 4000]
```
Comment (HajimeKawahara) [inline tests/xs/opacity_Fe_test_gamma_KA4.py:49]: In the current version, this test has 40 items and it takes very long time. But it's just a unit test. can you remove the number of items? Probably, only 1 parameter set is enough for the purpose of the unit test?

Comment (HajimeKawahara) [reply tests/xs/opacity_Fe_test_gamma_KA4.py:49]: But anyway, it worked! (though I cannot wait for the all items....)

``` sh
shirochan:~/exojax/tests/xs(metalline)>pytest opacity_Fe_test.py
================================================= test session starts ==================================================
platform linux -- Python 3.8.8, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/kawahara/exojax
plugins: anyio-2.2.0, asdf-2.8.1
collected 40 items                                                                                                     

opacity_Fe_test.py .....................
```

Comment (HajimeKawahara) [reply tests/xs/opacity_Fe_test_gamma_KA4.py:49]: Oh, sorry this was for "opacity_Fe_test.py" not "opacity_Fe_test_gamma_KA4.py" but the same comments for those unit tests....

File: tests/xs/opacity_Fe_test.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+#$ cp [user_name_at_VALD].[request_number_at_VALD].gz vald2600.gz
+adbFe = moldb.AdbVald(filepath_VALD3, nus4LL, Irwin=pf_Irwin)
+
+Amol=np.float64( adbFe.atomicmass[0] ) #atomic mass [u]
+ionE=np.float64( adbFe.ionE[0] ) #ionization energy [eV]
+nu0=adbFe.nu_lines
+
+#-------
+
+@pytest.mark.parametrize("T", [81, 200, 493, 1215, 2217, 2995, 3250, 4000]) #[81, 110, 148, 200, 270, 365, 493, 666, 900, 1215, 1641, 2000, 2217, 2500, 2750, 2995, 3250, 3500, 3750, 4000]
```
Comment (HajimeKawahara) [inline tests/xs/opacity_Fe_test.py:49]: See the comments in "opacity_Fe_test_gamma_uns.py".

File: tests/xs/opacity_Fe_test.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
+"""test for opacity calculation of metal lines
```
Comment (HajimeKawahara) [inline tests/xs/opacity_Fe_test.py:1]: In the current version,  there are 5 unit tests for Fe opacity:
```
opacity_Fe_test.py
opacity_Fe_test_gamma_KA3.py
opacity_Fe_test_gamma_KA3s.py
opacity_Fe_test_gamma_KA4.py
opacity_Fe_test_gamma_uns.py
```
It looks a lot. Can you combine these tests to the one unit test?
Instead, I recommend to reduce the number of items for each test to save the time.

**Pull request merged**